newrelic/newrelic-diagnostics-cli
# Release workflow

# Warning Note
Before you start reviewing the section "Files changed", you can use my filtered version of "files changed": https://github.com/newrelic/newrelic-diagnostics-cli/pull/12/files/37b502bdaf41671b4ab94fdbc5ad507d651d2195..10b42ae3b2af4e6c70659a7eb227ada9a7d2a33c

Without the filter, you would see all the changes that got carried from my previous PR/Branch into this new PR.

# Issue
Prior to open sourcing, our deploy and release process was integrated with Jenkins and pdx.artifactory.
Jenkins configuration: https://support-build.pdx.vm.datanerd.us/me/my-views/view/all/job/NR-Diagnostics--master--build-Upload-Download.Newrelic.com/configure
This internal configuration and tools are no longer available to us in our open source project.

# Goals
-Replace Jenkins with Github actions for our release configuration and process
-Use Github Secrets for storing and pulling AWS credentials that go into uploading artifacts to download.newrelic.com
-Simplify the release process so that all we need to do is push up a new tag to our repo and the release gets cut
-After applying changes, be able to find the new release cut in downloads.newrelic.com

# Implementation Details
The new release workflow is implemented here: https://github.com/newrelic/newrelic-diagnostics-cli/blob/10b42ae3b2af4e6c70659a7eb227ada9a7d2a33c/.github/workflows/release.yml

This workflow will be triggered if we push a git tag that is prepended with the words "build/"

In order to publish the new nrdiag release in the downloads.newrelic.com site, we need to upload files to the AWS bucket that stores the artifacts published on that site. The AWS credentials are being pulled from https://github.com/newrelic/newrelic-diagnostics-cli/settings/secrets

To see the latest release workflow successfully uploading to download.newrelic.com, you can look at these logs:
https://github.com/newrelic/newrelic-diagnostics-cli/runs/1321869840?check_suite_focus=true

The real path to release a new nrdiag version on the downloads site is: https://download.newrelic.com/nrdiag
But you'll notice that I uploaded to: https://download.newrelic.com/nrdiag/test/ . I temporarily created that /test/ path so I would not have to release a "real" new version of nrdiag ^_^

# How to Test
If you wish to see the workflow running live, you'll have to:

1. change and commit the s3 path in here to point `s3://${S3_BUCKET}/nrdiag/test/`
https://github.com/newrelic/newrelic-diagnostics-cli/blob/10b42ae3b2af4e6c70659a7eb227ada9a7d2a33c/scripts/upload.sh#L42-L44
That way you won't release an "official" new version 😬

2. create a new tag in your local repo and push it to the remote repo. Something like this:
`git tag build/testing`
`git push origin build/testing`

You should see the release workflow you just invoked right here: https://github.com/newrelic/newrelic-diagnostics-cli/actions?query=workflow%3ARelease

You can also verify the existence of the newly uploaded files that come with the release in the downloads s3 bucket by running this cmd in your terminal:
`[YOUR-AWSCREDENTIALS-HERE] aws s3 ls s3://bucketname/nrdiag/test/`

Note: The AWS credentials have been shared in LastPass. It's under the name `Madhatter upload`
